File: ALM-Model-Lab/research/bootstrap.py
"""SOFR OIS zero-coupon bootstrap.

Replaces the prototype's `bootstrapZeroCurve` (lines 422-447 of the Gemini
source) with conventions that match USD SOFR OIS practice.

Conventions
-----------
* Day count: ACT/360 for cash, ACT/360 for swap fixed (USD SOFR OIS standard).
  The prototype used integer-year accrual, undercounting by 1.4% (5 days / 360);
  this changes long-end discount factors by 1-3 bp.
* Cash leg: discount factor = 1 / (1 + r * accrual). The prototype used
  simple compounding too, so this is unchanged in form, only in accrual.
* Swap leg: annual fixed schedule, par swap rate priced under standard
  formula PV_fix = K * sum(tau_i * DF_i), PV_float = 1 - DF(0, T_n).
  Setting equal gives the bootstrap equation, solved by Newton-Raphson with
  closed-form derivative dPV_fix/dz_n.
* Interpolation between zero-rate nodes: linear in zero rate.

Outputs
-------
ZeroCurve: a pair of numpy arrays (t_years, z_rate) plus a callable for
log-linear-in-DF interpolation, suitable as input to HW and BGM pricers.

References
----------
- Andersen-Piterbarg Vol I §6 (curve construction).
- Brigo-Mercurio §1.2 (yield-curve conventions).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable
import os

import numpy as np
from scipy.optimize import brentq

from market_data import MarketSnapshot, curve_arrays


@dataclass(frozen=True)
class ZeroCurve:
    """A bootstrapped continuously compounded zero-coupon curve.

    Attributes
    ----------
    t : np.ndarray
        Tenors in years for each node.
    z : np.ndarray
        Continuously compounded zero rates at each node.
    """
    t: np.ndarray
    z: np.ndarray

    def discount_factor(self, time: float | np.ndarray) -> float | np.ndarray:
        """DF(0, time) by linear interpolation in the zero rate.

        Off the curve edges, holds the endpoint rate flat (consistent with
        the prototype's interpLinear behaviour).
        """
        z_interp = np.interp(time, self.t, self.z)
        return np.exp(-z_interp * np.asarray(time))

    def zero_rate(self, time: float | np.ndarray) -> float | np.ndarray:
        return np.interp(time, self.t, self.z)

    def forward_rate(self, t1: float, t2: float) -> float:
        """Simply-compounded forward rate between t1 and t2."""
        if t2 <= t1:
            raise ValueError(f"t2={t2} must exceed t1={t1}")
        return (self.discount_factor(t1) / self.discount_factor(t2) - 1.0) / (t2 - t1)


def _annual_fix_schedule(t_n: float) -> np.ndarray:
    """Annual fixed-leg payment dates up to t_n (inclusive). ACT/360 fraction."""
    # USD SOFR OIS: annual fixed coupons. Schedule is years 1, 2, ..., n where
    # n = round(t_n). For a 4Y swap quoted at t_years=4.0 we expect [1, 2, 3, 4].
    n = int(round(t_n))
    if abs(n - t_n) > 1e-6:
        raise ValueError(f"non-integer maturity {t_n} not supported by annual schedule")
    return np.arange(1, n + 1, dtype=float)


def bootstrap_zero_curve(snapshot: MarketSnapshot) -> ZeroCurve:
    """Bootstrap a continuously compounded zero curve from SOFR OIS quotes.

    Cash quotes (t <= 1.0) are converted directly via simple compounding.
    Swap quotes (t > 1.0) are bootstrapped sequentially by Newton-style
    root finding on the par-swap equation, given prior nodes already solved.
    """
    t_in, r_in = curve_arrays(snapshot)
    if not np.all(np.diff(t_in) > 0):
        raise ValueError("curve quotes must be sorted by t_years strictly ascending")

    t_nodes: list[float] = []
    z_nodes: list[float] = []

    for t_i, r_i, q in zip(t_in, r_in, snapshot.curve_quotes):
        if q.instrument_type == "CASH" or t_i <= 1.0:
            # Simple compounded cash conversion.
            df = 1.0 / (1.0 + r_i * t_i)
            z = -np.log(df) / t_i
            t_nodes.append(t_i)
            z_nodes.append(z)
            continue

        # SWAP > 1Y: solve for z_n such that PV_fix(K=r_i; z_n) = PV_float(z_n).
        prior_t = np.array(t_nodes)
        prior_z = np.array(z_nodes)
        schedule = _annual_fix_schedule(t_i)

        def par_residual(z_n: float) -> float:
            # Build a candidate curve including the trial z_n.
            tt = np.append(prior_t, t_i)
            zz = np.append(prior_z, z_n)
            # Discount factors at coupon dates by linear interpolation in z.
            z_at_coupons = np.interp(schedule, tt, zz)
            df_coupons = np.exp(-z_at_coupons * schedule)
            # ACT/360 accrual fraction between consecutive coupon dates.
            # First coupon: prior point is t=0 (DF=1); accrual = year 1.
            tau = np.diff(np.concatenate(([0.0], schedule))) * 365.0 / 360.0
            pv_fixed = r_i * np.sum(tau * df_coupons)
            df_n = np.exp(-z_n * t_i)
            pv_float = 1.0 - df_n
            return pv_float - pv_fixed

        # Bracket: z_n bounded loosely by the surrounding rates.
        lo = max(prior_z[-1] - 0.05, 1e-5)
        hi = prior_z[-1] + 0.10
        try:
            z_solved = brentq(par_residual, lo, hi, xtol=1e-12, rtol=1e-12, maxiter=200)
        except ValueError:
            # Fall back to a wider bracket if the initial one missed the root.
            z_solved = brentq(par_residual, 1e-5, 0.30, xtol=1e-12, rtol=1e-12, maxiter=200)

        t_nodes.append(t_i)
        z_nodes.append(z_solved)

    return ZeroCurve(t=np.array(t_nodes), z=np.array(z_nodes))


def make_curve_callable(curve: ZeroCurve) -> Callable[[float], float]:
    """Return a callable DF(0, t) -> float for downstream pricers."""
    def _df(t: float) -> float:
        return float(curve.discount_factor(t))
    return _df


if __name__ == "__main__":
    from pathlib import Path
    from market_data import load_market_snapshot

    snap = load_market_snapshot(
        Path(__file__).parent / "data" / os.environ.get("MARKET_JSON", "market_2025-09-30.json")
    )
    curve = bootstrap_zero_curve(snap)
    print(f"Bootstrapped {len(curve.t)} nodes from {snap.calibration_date}:")
    for t, z in zip(curve.t, curve.z):
        df = float(curve.discount_factor(t))
        print(f"  t={t:8.4f}y  z={z:8.6f}  DF={df:8.6f}")

    # Round-trip check: bootstrap then reprice each swap and report bp error.
    print("\nRepricing check (bp error on each input swap):")
    for q in snap.curve_quotes:
        if q.instrument_type != "SWAP" or q.t_years <= 1.0:
            continue
        schedule = _annual_fix_schedule(q.t_years)
        z_at = curve.zero_rate(schedule)
        df_at = np.exp(-z_at * schedule)
        tau = np.diff(np.concatenate(([0.0], schedule))) * 365.0 / 360.0
        pv_fix_unit = float(np.sum(tau * df_at))
        df_n = float(curve.discount_factor(q.t_years))
        par_implied = (1.0 - df_n) / pv_fix_unit
        err_bp = (par_implied - q.rate) * 1e4
        print(f"  {q.term:>4s}  market={q.rate*100:6.4f}%  implied={par_implied*100:6.4f}%  err={err_bp:+.4f} bp")

File: ALM-Model-Lab/src/math/rates/bootstrap.ts
/**
 * SOFR OIS zero-coupon bootstrap.
 *
 * Port of `research/bootstrap.py`. Conventions match USD SOFR OIS practice:
 * - Day count: ACT/360 (annual fixed coupon = 365/360 fraction).
 * - Cash leg: simple compounding DF = 1 / (1 + r * t).
 * - Swap leg: annual fixed schedule, par equation solved by Brent root finding.
 * - Interpolation between zero-rate nodes: linear in zero rate.
 *
 * References
 * - Andersen-Piterbarg Vol I Sec 6 (curve construction).
 */

import type { MarketSnapshot } from "./marketData";
import { brentq } from "./rootFind";

export interface ZeroCurve {
  readonly t: number[]; // tenors in years
  readonly z: number[]; // continuously compounded zero rates
  zeroRate(time: number): number;
  discountFactor(time: number): number;
  /**
   * Simply-compounded forward rate over [t1, t2]:
   *   f = (P(t1)/P(t2) - 1) / (t2 - t1)
   * Money-market convention. Correct for τ ≤ 1Y (caps, FTP, SOFR fix
   * derivation). For τ > 1Y, use forwardSwapRate instead — simple
   * compounding overstates long-tenor rates by hundreds of bps.
   */
  forwardRate(t1: number, t2: number): number;
  /**
   * Forward par swap rate for an annual ACT/360 fixed schedule:
   *   For τ < 1Y: simple-compounded forward (single-payment OIS convention).
   *   For τ ≥ 1Y, integer: S = (P(t1) - P(t2)) / (Σ_k (365/360) · P(t1+k)).
   * This matches the convention used by the bootstrap (par OIS) and by
   * buildSwap (BGM/SABR-on-swaptions). Use this for any swap-rate driver
   * (mortgage prepayment, long-tenor forward display, etc.).
   */
  forwardSwapRate(t1: number, t2: number): number;
}

/** Annual fixed coupon fraction for SOFR OIS (ACT/360). */
const ACT_360_ANNUAL = 365 / 360;

function lerp(xs: number[], ys: number[], x: number): number {
  if (xs.length === 0) return 0;
  if (x <= xs[0]) return ys[0];
  if (x >= xs[xs.length - 1]) return ys[xs.length - 1];
  // binary search would be O(log n); array sizes are tiny here
  for (let i = 0; i < xs.length - 1; i++) {
    if (x >= xs[i] && x <= xs[i + 1]) {
      const w = (x - xs[i]) / (xs[i + 1] - xs[i]);
      return ys[i] * (1 - w) + ys[i + 1] * w;
    }
  }
  return ys[ys.length - 1];
}

function makeCurve(t: number[], z: number[]): ZeroCurve {
  return {
    t,
    z,
    zeroRate(time: number): number {
      return lerp(t, z, time);
    },
    discountFactor(time: number): number {
      return Math.exp(-this.zeroRate(time) * time);
    },
    forwardRate(t1: number, t2: number): number {
      if (t2 <= t1) {
        throw new Error(`forwardRate: t2=${t2} must exceed t1=${t1}`);
      }
      return (this.discountFactor(t1) / this.discountFactor(t2) - 1) / (t2 - t1);
    },
    forwardSwapRate(t1: number, t2: number): number {
      if (t2 <= t1) {
        throw new Error(`forwardSwapRate: t2=${t2} must exceed t1=${t1}`);
      }
      const tau = t2 - t1;
      // Boundary matches the bootstrap:
      //   τ ≤ 1Y → cash / single-payment OIS, simple-compounded
      //            (no ACT/360 day-count scaling — same convention the
      //            bootstrap uses for q.tYears ≤ 1.0, so round-trip is exact).
      //   τ > 1Y → annual ACT/360 fixed schedule, par swap rate
      //            (same as bgmPricing.buildSwap).
      if (tau <= 1.0 + 1e-9) {
        return (this.discountFactor(t1) / this.discountFactor(t2) - 1) / tau;
      }
      const n = Math.round(tau);
      let annuity = 0;
      for (let k = 1; k <= n; k++) {
        annuity += ACT_360_ANNUAL * this.discountFactor(t1 + k);
      }
      const P1 = this.discountFactor(t1);
      const Pn = this.discountFactor(t1 + n);
      return (P1 - Pn) / annuity;
    },
  };
}

function annualFixSchedule(tN: number): number[] {
  const n = Math.round(tN);
  if (Math.abs(n - tN) > 1e-6) {
    throw new Error(`non-integer maturity ${tN} not supported by annual schedule`);
  }
  const sched: number[] = [];
  for (let i = 1; i <= n; i++) sched.push(i);
  return sched;
}

export function bootstrapZeroCurve(snapshot: MarketSnapshot): ZeroCurve {
  const quotes = [...snapshot.curveQuotes];
  for (let i = 0; i < quotes.length - 1; i++) {
    if (!(quotes[i].tYears < quotes[i + 1].tYears)) {
      throw new Error("curve quotes must be strictly ascending in t_years");
    }
  }

  const tNodes: number[] = [];
  const zNodes: number[] = [];

  for (const q of quotes) {
    if (q.instrumentType === "CASH" || q.tYears <= 1.0) {
      const df = 1.0 / (1.0 + q.rate * q.tYears);
      const z = -Math.log(df) / q.tYears;
      tNodes.push(q.tYears);
      zNodes.push(z);
      continue;
    }

    const priorT = [...tNodes];
    const priorZ = [...zNodes];
    const schedule = annualFixSchedule(q.tYears);

    const parResidual = (zN: number): number => {
      const tt = [...priorT, q.tYears];
      const zz = [...priorZ, zN];
      let pvFixed = 0;
      let prevDate = 0;
      for (const couponDate of schedule) {
        const zAt = lerp(tt, zz, couponDate);
        const dfAt = Math.exp(-zAt * couponDate);
        const tau = (couponDate - prevDate) * (365 / 360);
        pvFixed += q.rate * tau * dfAt;
        prevDate = couponDate;
      }
      const dfN = Math.exp(-zN * q.tYears);
      const pvFloat = 1.0 - dfN;
      return pvFloat - pvFixed;
    };

    const lo = Math.max(priorZ[priorZ.length - 1] - 0.05, 1e-5);
    const hi = priorZ[priorZ.length - 1] + 0.10;
    let zSolved: number;
    try {
      zSolved = brentq(parResidual, lo, hi);
    } catch {
      zSolved = brentq(parResidual, 1e-5, 0.30);
    }
    tNodes.push(q.tYears);
    zNodes.push(zSolved);
  }

  return makeCurve(tNodes, zNodes);
}

/**
 * Parallel-shifted copy of a zero curve: every node's zero rate moves by
 * shockBp. Interpolation is linear in zero rate, so zeroRate(t) shifts by
 * exactly s at every t and discount factors scale by exp(-s·t). This is the
 * EVE-style curve shock (discounting side); ShockedPath in seg.ts shifts only
 * the projection path (cashflow side).
 */
export function shockCurve(curve: ZeroCurve, shockBp: number): ZeroCurve {
  const s = shockBp / 1e4;
  return makeCurve([...curve.t], curve.z.map((z) => z + s));
}

export function _annualFixSchedule(tN: number): number[] {
  return annualFixSchedule(tN);
}

export function _curveLerp(t: number[], z: number[], x: number): number {
  return lerp(t, z, x);
}
